### DISCUSSION TITLE
First 1 or 2 tool calls succeed with Ollama, subsequent tool calls fail with exposed <tool_call> in assistant response

### DISCUSSION THREAD

**Discussion by dto** [ahyatt/llm#94] [Q&A]
Instead of the correct Emacs Lisp functions being invoked by llm, after a few tries I start to get stuff like this in the assistant message instead of as a proper tool call:

```
<tool_call>
{"name": "systemName", "arguments": null}
</tool_call>
```

Seems to happen no matter which tools enabled model i use.

**Comment by ahyatt** [comment]
It certainly possible that this is a bug of some sort.  Is this in a conversation (using the same prompt over and over) or with separate prompts?

**Reply by dto** [reply]
This is in a conversation. I can provide more debug info, I'll make another post on this thread.

**Reply by dto** [reply]
I'm reading up here on the Qwen model I'm using: https://qwen.readthedocs.io/en/latest/framework/function_call.html where it says "As before, chances are that there are corner cases where tool calls are generated but they are malformed and cannot be parsed. For production code, we should try parsing by ourselves."  Which I am doing now, to catch tool calls that aren't caught by Emacs and/or Ollama. I think the error may be on the Ollama side because the role:assistant is set already when Emacs gets the response (I think)

**Reply by ahyatt** [reply]
Multiple conversational rounds with function calling is always a bit tricky because every provider has their own pecularities with it. I'll look into this in the next few days.

**Comment by dto** [comment]
Thank you! I'll keep my eyes peeled, and i'll keep trying to debug. Might
have to put some print statements into PLZ
